src/Sarif.Sarifer/BackgroundAnalyzerBase.cs

```diff
@@ -0,0 +1,72 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel.Composition;
+using System.IO;
+using System.Text;
+
+using Microsoft.VisualStudio.Shell;
+
+using Task = System.Threading.Tasks.Task;
+
+namespace Microsoft.CodeAnalysis.Sarif.Sarifer
+{
+    /// <summary>
+    /// Base class for background analyzers.
+    /// </summary>
+    /// <remarks>
+    /// This class invokes the analysis (implemented in derived classes), and sends the resulting
+    /// <see cref="SarifLog"/> to all exported implementations of <see cref="IBackgroundAnalysisSink"/>.
+    /// Derived classes need only override <see cref="CreateSarifLog(string)"/>.
+    /// </remarks>
+    public abstract class BackgroundAnalyzerBase : IBackgroundAnalyzer
+    {
+#pragma warning disable CS0649 // Filled in by MEF
+#pragma warning disable IDE0044 // Assigned by MEF
+        [ImportMany]
+        private IEnumerable<IBackgroundAnalysisSink> sinks;
+#pragma warning restore IDE0044
+#pragma warning restore CS0649
+
+        /// <inheritdoc/>
+        public async Task StartAnalysisAsync(string path, string text)
+        {
+            text = text ?? throw new ArgumentNullException(nameof(text));
+
+            using (Stream stream = new MemoryStream())
+            using (TextWriter writer = new StreamWriter(stream, Encoding.UTF8))
+            {
+                CreateSarifLog(path, text, writer);
+                await writer.FlushAsync().ConfigureAwait(continueOnCapturedContext: false);
+
+                foreach (IBackgroundAnalysisSink sink in sinks)
+                {
+                    stream.Seek(0L, SeekOrigin.Begin);
+                    await sink.ReceiveAsync(stream).ConfigureAwait(continueOnCapturedContext: false);
```

Comment:
>sink.ReceiveAsync(stream) [](start = 26, length = 25)

Hmm.. seeking the stream to zero won't necessary help if an implementation of ReceiveAsync does something on a completely background task or, for example, another thread.

You should consider new StreamReader(memoryStream) and pass StreamReader.BaseStream to ReceiveAsync.

That will also allow you to start all the ReceiveAsync in parallel and return Task.All()